data/rule/rule.go

```diff
@@ -0,0 +1,117 @@
+// Package rule contains rule-based heuristic implementations.
+// It is used in the generated code in content.go for disambiguation of languages
+// with colliding extensions based on regexps from Linguist data.
+package rule
+
+import (
+	"regexp"
+)
+
+// Heuristic consist of a number of rules where each, if matches,
+// identifes content as belonging to a programming language(s).
+type Heuristic interface {
+	Matcher
+	Languages() []string
+}
+
+// Matcher checks if the data matches (number of) pattern.
+// Every rule below implements this interface: a rule is matcher that identifies
+// given programming language(s) in case of the match.
+type Matcher interface {
+	Match(data []byte) bool
+}
+
+// languages struct incapsulate data common to every Matcher: all languages
+// that it identifies.
+type languages struct {
+	langs []string
+}
+
+// Languages returns all languages, identified by this Matcher.
+func (l *languages) Languages() []string {
```

Comment:
Need this be passed by pointer? I suggest passing by value so there is no extra indirection required—the slice inside is already a reference value.

Reply:
(then there is no possibility of a nil pointer indirection from improper use, either—the compiler won't allow it)

Reply:
I agree and appreciate the suggestions. Biggest impact on not having this as a pointer would be such change in the generated code

```
                        rule.Not(
-                               nil,
+                               rule.MatchingLanguages(""),
                                regexp.MustCompile(`(?m)/\*`),
                        ),
```

Reply:
Adressed in ec54891